Issue #2: Add support for group authorizations with Apache 2.4.x

esabol opened the pull request:
This PR addresses issue #2 by adding support for group authorizations when used with Apache 2.4.x. It is currently in a separate Perl module, Apache2_4::AuthCookieDBI. It seems cleaner that way to me (because the Apache2::Consts are different under Apache 2.4.x), but I'm open to possibly combining the code into Apache2::AuthCookieDBI. Alternatively, I imagine much of the group code could be refactored to avoid duplication. Let me know if you want any changes made.

Apache2.4.x calls the authorization provider multiple times per request. In my testing, I was seeing 4+ calls per request for URLs that were directories. Apache was authz'ing the directory and then the index.html in that directory. Anyway, to eliminate multiple database queries for the same request, I was wondering if checking for the presence of `$r->subprocess_env('AUTH_COOKIE_DBI_GROUP')` and using that instead of querying the database again was secure? It all seems to work in my testing and it drastically cuts down on repeated database queries, but I'm worried that might be insecure. What do you think?

esabol committed:
Add support for group authorizations with Apache 2.4.x

Add support for group authorizations with Apache 2.4.x - tweak

Add support for group authorizations with Apache 2.4.x - tweak 2

Add support for group authorizations with Apache 2.4.x - tweak 3

Add support for group authorizations with Apache 2.4.x - updated Changes

Add support for group authorizations with Apache 2.4.x - tweak 5

matisse commented:
I’m traveling and will review this next week.

esabol commented:
Ping!

matisse commented:
Hi Ed - I have been back 2 weeks and simply haven't been able to get to it. It is on my list for this weekend.

I apologize for being so slow.

-------------------------------------------------------
Matisse Enzer <matisse@matisse.net>




> On Nov 5, 2019, at 3:23 PM, Ed Sabol <notifications@github.com> wrote:
> 
> Ping!
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/matisse/Apache-AuthCookieDBI/pull/6?email_source=notifications&email_token=AAEBFLJ2E5YPCQCTXHCIYD3QSH57RA5CNFSM4JDYSEXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDEWGFQ#issuecomment-550069014>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAEBFLNNSBR4JEIOGPRMSXLQSH57RANCNFSM4JDYSEXA>.
>

matisse commented:
Ed: If you have no more changes to make I'll merge, then update the actual release date in master just before i submit to CPAN.

esabol commented:
If you don’t want any changes (code refactorization? tests?) and you aren’t concerned about the security of using `$r->subprocess_env('AUTH_COOKIE_DBI_GROUP')` on subsequent requests to eliminate duplicate queries, then go ahead.

matisse commented:
It would be great if there were tests, and even better if a way could be found to have a single module instead if forking the code for 2 v 2.4, perhaps but I don’t want to block your adoption on my slowness, as I have some extra intense stuff going on that makes me less responsive currently.

> On Nov 9, 2019, at 11:59 AM, Ed Sabol <notifications@github.com> wrote:
> 
> ﻿
> If you don’t want any changes (code refactorization? tests?) and you aren’t concerned about the security of using $r->subprocess_env('AUTH_COOKIE_DBI_GROUP') on subsequent requests to eliminate duplicate queries, then go ahead.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

esabol commented:
I don’t see a need to rush out a release. AuthCookieDBI hasn’t worked with Apache 2.4.x for years now. And, if you release a new version with a separate `Apache2_4::AuthCookieDBI` module, there’s probably no going back. That’s what happened with `Apache2::AuthCookie`. It still ships with a `Apache2_4::AuthCookie` even though it’s not needed at this point.

That said, I’m not sure how to combine it into one module. I thought I had come across another module which combined code for different Apache versions, but I’ve lost it. If you have any suggestions, let me know.

matisse commented:
Ok I will give it some thought and reply tomorrow.

> On Nov 9, 2019, at 2:04 PM, Ed Sabol <notifications@github.com> wrote:
> 
> ﻿
> I don’t see a need to rush out a release. AuthCookieDBI hasn’t worked with Apache 2.4.x for years now. And, if you release a new version with a separate Apache2_4::AuthCookieDBI module, there’s probably no going back. That’s what happened with Apache2::AuthCookie. It still ships with a Apache2_4::AuthCookie even though it’s not needed at this point.
> 
> That said, I’m not sure how to combine it into one module. I thought I had come across another module which combined code for different Apache versions, but I’ve lost it. If you have any suggestions, let me know.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

matisse commented on lib/Apache2_4/AuthCookieDBI.pm:
```diff
@@ -0,0 +1,225 @@
+    # the authorization.
+    if ( my $cached_group = $r->subprocess_env('AUTH_COOKIE_DBI_GROUP') ) {
+	$r->server->log_error("${class}\tfound cached group $cached_group") if ($debug >= 3);
+	foreach my $group (@groups) {
```
I think we can skip the foreach loop here, and move this check up to the top of the method, potentially returning even earlier.

My thinking is that if `AUTH_COOKIE_DBI_GROUP` subprocess_env is set then we have already returned `AUTHZ_GRANTED` at least once, and so we might as well keep returning `AUTHZ_GRANTED` as early as possible in this method.

Does that make sense?

esabol replied:
No, I disagree. I don't think just checking for the existence of AUTH_COOKIE_DBI_GROUP is secure. I think the location being authorized could have different requirements from a previous request in the authorization request chain. For example, let's say a user goes to the url https://myserver.com/protected. In my tests, Apache will call the authz provider four times: once with a blank user (AUTHZ_DENIED_NO_USER required for that), once for /protected, once for /protected/, and finally for /protected/index.html. I can imagine a scenario where some of those last three locations might have different group requirements. It's unlikely in practice, but I don't think it's strictly correct to skip the check entirely. And I don't see how this check could be moved any earlier in the method.

matisse commented on lib/Apache2_4/AuthCookieDBI.pm:
```diff
@@ -0,0 +1,225 @@
+for L<Apache2::AuthCookieDBI> group-based authorizations.
+
+=head1 SYNOPSIS
+
```
Perhaps note that this is a sub-class of `Apache2::AuthCookieDBI`

esabol replied:
Sure. Would putting that note under the DESCRIPTION be acceptable? Or would you prefer it to be under SYNOPSIS? Or under NAME, like so:

```Perl
=head1 NAME                                                                                              
                                                                                                         
Apache2_4::AuthCookieDBI - A subclass of L<Apache2::AuthCookieDBI>                                       
that implements group-based authorization for Apache 2.4.x.                                              
```

matisse replied:
I think `DESCRIPTION` is actually better, so go with that.

matisse commented on lib/Apache2_4/AuthCookieDBI.pm:
```diff
@@ -0,0 +1,225 @@
+
+# ABSTRACT: Perl group authorization for Apache 2.4.x
+
+use strict;
```
Can we add `use warnings;` I'd really prefer it to be added unless it breaks something we cannot work around.

esabol replied:
Yes, of course.

matisse reviewed (changes requested):
OK Ed I think for now using a single file as subclass probably was the best choice, but i did add some comments/requests.

I know the dirt has virtually no tests, and that's my fault. the libraries in `t/mock-libs` are supposed to make testing easier, but a set of tests has not been added. I think I started on them and got sidetracked somewhere, I actually don't remember now what happened :-(

esabol commented:
Do you think it makes sense to refactor the following code into a method that can be used by both `Apache2::AuthCookieDBI` and `Apache2_4::AuthCookieDBI`?

```Perl
sub _prepare_group_query {
    my ( $class, $r, $dbh ) = @_;

    # Get the configuration information.                                                                 
    my %c = $class->_dbi_config_vars($r);
    my $DBI_GroupUserField = $dbh->quote_identifier($c{'DBI_GroupUserField'});
    my $DBI_GroupsTable = $dbh->quote_identifier($c{'DBI_GroupsTable'});
    my $DBI_GroupField = $dbh->quote_identifier($c{'DBI_GroupField'});

    my $sth = $dbh->prepare_cached( <<"EOS" );                                                           
SELECT $DBI_GroupUserField                                                                               
FROM $DBI_GroupsTable                                                                                    
WHERE $DBI_GroupField = ?                                                                                
AND $DBI_GroupUserField = ?                                                                              
EOS                                                                                                      

    return $sth;
}
```

esabol commented:
Also, can I move

```Perl
use constant LOG_TYPE_AUTHZ   => 'authorization';
```

to `Apache2::AuthCookieDBI` and have the `group()` method there use that for the call to `logger()` instead of `LOG_TYPE_AUTH`?

esabol commented on lib/Apache2_4/AuthCookieDBI.pm:
```diff
@@ -0,0 +1,225 @@
+        return Apache2::Const::AUTHZ_DENIED;
+    }
+
+    my @groups = split( Apache2::AuthCookieDBI::WHITESPACE_REGEX, $groups );
```
Should I use `$class->WHITESPACE_REGEX` here instead?

matisse commented:
> Do you think it makes sense to refactor the following code into a method that can be used by both `Apache2::AuthCookieDBI` and `Apache2_4::AuthCookieDBI`?
> 
> ```perl
> sub _prepare_group_query {
>     my ( $class, $r, $dbh ) = @_;
> 
>     # Get the configuration information.                                                                 
>     my %c = $class->_dbi_config_vars($r);
>     my $DBI_GroupUserField = $dbh->quote_identifier($c{'DBI_GroupUserField'});
>     my $DBI_GroupsTable = $dbh->quote_identifier($c{'DBI_GroupsTable'});
>     my $DBI_GroupField = $dbh->quote_identifier($c{'DBI_GroupField'});
> 
>     my $sth = $dbh->prepare_cached( <<"EOS" );                                                           
> SELECT $DBI_GroupUserField                                                                               
> FROM $DBI_GroupsTable                                                                                    
> WHERE $DBI_GroupField = ?                                                                                
> AND $DBI_GroupUserField = ?                                                                              
> EOS                                                                                                      
> 
>     return $sth;
> }
> ```

Yes, and given your (correct I think) comment on the need to re-check the groups for each auth call I think the `foreach` loop that check the group could also be moved into the new base-class method (which, but, have that method return a simple boolean instead of an Apache constant, and then the caller uses the correct Apache constant depending on whether the caller is Apache2_4 or not.

Maybe the whole `group` method could become a private method `_group` in the base class that returns a simple boolean, and the actual `group` methods in two classes are the ones that use the differing constants?

How does that sound?
(I have to run to work now ... rushing on this so maybe I missed something.)

esabol commented:
> Maybe the whole group method could become a private method _group in the base class that returns a simple boolean, and the actual group methods in two classes are the ones that use the differing constants?
>
> How does that sound?

Yeah, that makes sense. I'll try on it tomorrow when I'm at work.

May I modify `_dbi_connect()` to take an optional additional argument containing the hash with the configuration information? I've noticed that `$class->_dbi_config_vars($r)` is called multiple times by several methods, typically by `_dbi_connect()` and the method which calls `_dbi_connect()`. I haven't profiled it, but this seems rather inefficient.

I'm thinking something like this:

```Perl
sub _dbi_connect {
    my ( $class, $r, $config_hash ) = @_;
...
    my %c = ref($config_hash) eq 'HASH' ? %$config_hash : $class->_dbi_config_vars($r);
...
```
Or just
```Perl
    my %c = $config_hash ? %$config_hash : $class->_dbi_config_vars($r);
```
if you think we can safely skip checking that it's a hash reference before using it.

matisse commented:
Yes that’s fine.

esabol committed:
Add support for group authorizations with Apache 2.4.x - refactor code

esabol commented:
Sorry, I was swamped at work all week and just got to this today.

I just pushed a bunch of changes to the PR. I still need to test these changes on our server, but please let me know what you think of the direction.

esabol committed:
Add support for group authorizations with Apache 2.4.x - refactor code (minor tweaks)

esabol commented:
> I added comments for 2 tiny things, otherwise looks OK to me.

I have pushed another commit which addresses these two things.

esabol commented:
`$r` probably doesn't need to be passed to `_prepare_group_method()`. It's only used if `$config_hash` isn't passed as an argument, which it is in both cases where that method is called. So I'll probably remove `$r` from the list of arguments unless you think it might be useful to someone subclassing AuthCookieDBI?

matisse commented:
> `$r` probably doesn't need to be passed to `_prepare_group_method()`. It's only used if `$config_hash` isn't passed as an argument, which it is in both cases where that method is called. So I'll probably remove `$r` from the list of arguments unless you think it might be useful to someone subclassing AuthCookieDBI?

Here are two thoughts on this: One is that if it is not needed in this private method then it is better (I think) to not have it. Second is that this is a private method so no-one should override it in a subclass (yes I know they might).

esabol committed:
Add support for group authorizations with Apache 2.4.x - refactor code (more tweaks)

esabol committed:
Add support for group authorizations with Apache 2.4.x - refactor code (updated Changes)

esabol commented:
OK, I have removed `$r` as an argument to `_prepare_group_query()`. I'll get back to you soon with the results of my testing.

esabol committed:
Add support for group authorizations with Apache 2.4.x - improved error handling, fixed typos

esabol committed:
Add support for group authorizations with Apache 2.4.x - fix failing test (t/utils.t, 55)

esabol committed:
Add support for group authorizations with Apache 2.4.x - fix failing test (clean up)

esabol committed:
Add support for group authorizations with Apache 2.4.x - replaced dependency on Apache::AuthCookie::Util::is_blank with _is_empty

esabol commented:
I pushed a few more changes to improve error handling and fix some typos in `Apache2_4::AuthCookieDBI` that I introduced in the commit earlier this afternoon. I also removed the dependency on `Apache::AuthCookie::Util::is_blank` by calling `_is_empty()` instead.

I've completed testing all changes on our development server, and everything seems to be fine.

I can't think of any more code changes or improvements to make, so I'll call these my "final answer" unless you have any additional change requests.

Question: Do you recall why the `use 5.010_000;` was added in `Apache2/AuthCookieDBI.pm`? Older versions didn't have that. The code seems to work just fine on Perl 5.8.x. (Don't ask how I know this.)

Apache 2.4.x tests might be nice to have, but I don't have a good feel for how to implement that in the current test framework. What do you think?

esabol committed:
Add support for group authorizations with Apache 2.4.x - removed Id markers from files

esabol committed:
Add support for group authorizations with Apache 2.4.x - switched back to using is_blank

esabol commented:
> I can't think of any more code changes or improvements to make, so I'll call these my "final answer"...

Sorry, I was wrong. Turns out `_is_empty()` breaks things. Had to switch back to `is_blank()`. The semantics are slightly different.

matisse commented on lib/Apache2/AuthCookieDBI.pm:
```diff
@@ -557,15 +551,15 @@ sub _percent_decode {
         LOG_TYPE_SYSTEM, $r->uri );
 
-    my %c = $class->_dbi_config_vars($r);
+    my %c = $config_hash ? %$config_hash : $class->_dbi_config_vars($r);
```
In this method, and all private methods, that use the config hash, perhaps we do not need to check if `$config_hash` is true because all the callers of this method are in this class, so we do not need to guard against someone calling this method without passing `$config_hash` so maybe we should just do: `my %c = %$config_hash;` - what do you think?

esabol replied:
If it were a new method, sure, but I worry that someone out there has subclassed AuthCookieDBI. If so, they most likely have existing calls to `_dbi_connect()`, regardless of whether you consider it to be private or not, just as I did in `Apache2_4::AuthCookieDBI` when I developed that. Perl doesn’t enforce it, and we’re not using Moose or the like here. So I recommend keeping `$config_hash` as optional.

As for tests, I’ve tried and failed. I can’t get any Apache 2.4.x tests to work, even a simple `use_ok()`. I get “bareword” errors for the Apache 2.4.x AUTHZ_* constants. I think my server’s non-chrooted environment is messed up from too many Apache and mod_perl upgrades. It’s rather complicated, and I don’t have the permissions to change things. I could maybe try testing inside a Docker container, but I’d have to look around for one with Apache 2.4.41 and mod_perl 2.0.11....

Also, I think adding Apache 2.4.x-specific tests might change the dependency requirements on Apache::Test and maybe Apache itself. IIRC, Apache::Test’s minimum Apache version check requires Apache 2.0.40.

matisse reviewed (commented):
Added a question about $config_hash.
Also - do you want to try and add some test using the mock libs that are there?

matisse commented:
Ok that’s all reasonable. I’ll give another look and probably approve and merge tonight.

esabol commented:
Well, I found a Dockerfile that installed mod_perl, which I modified to install various modules and test Apache2_4::AuthCookieDBI... I got the same result:
```
Bareword "Apache2::Const::AUTHZ_DENIED_NO_USER" not allowed while "strict subs" in use at /var/tmp/Apache-AuthCookieDBI/blib/lib/Apache2_4/AuthCookieDBI.pm line 58.
Bareword "Apache2::Const::AUTHZ_DENIED" not allowed while "strict subs" in use at /var/tmp/Apache-AuthCookieDBI/blib/lib/Apache2_4/AuthCookieDBI.pm line 68.
Bareword "Apache2::Const::AUTHZ_GRANTED" not allowed while "strict subs" in use at /var/tmp/Apache-AuthCookieDBI/blib/lib/Apache2_4/AuthCookieDBI.pm line 79.
Bareword "Apache2::Const::AUTHZ_GENERAL_ERROR" not allowed while "strict subs" in use at /var/tmp/Apache-AuthCookieDBI/blib/lib/Apache2_4/AuthCookieDBI.pm line 91.
Bareword "Apache2::Const::AUTHZ_GENERAL_ERROR" not allowed while "strict subs" in use at /var/tmp/Apache-AuthCookieDBI/blib/lib/Apache2_4/AuthCookieDBI.pm line 93.
Bareword "Apache2::Const::AUTHZ_GRANTED" not allowed while "strict subs" in use at /var/tmp/Apache-AuthCookieDBI/blib/lib/Apache2_4/AuthCookieDBI.pm line 96.
Bareword "Apache2::Const::AUTHZ_DENIED" not allowed while "strict subs" in use at /var/tmp/Apache-AuthCookieDBI/blib/lib/Apache2_4/AuthCookieDBI.pm line 96.
```

I added '()' after all these Apache2::Consts, and the errors went away and `use_ok('Apache2_4::AuthCookieDBI');` passed ok.

This feels wrong though since you don't have to do this with other Apache2::Consts..... Do you have any insight into this?

esabol commented:
For reference, I used this Dockerfile:
```Dockerfile
FROM httpd:latest

ENV MOD_PERL_VERSION 2.0.11
ENV PERL_VERSION 5.28.0

ENV PATH /opt/perl-$PERL_VERSION/bin:$PATH

RUN apt-get update && \
    apt-get install -yq --no-install-recommends perl ca-certificates curl build-essential netbase && \
    curl -sfL https://raw.githubusercontent.com/tokuhirom/Perl-Build/master/perl-build | perl - $PERL_VERSION /opt/perl-$PERL_VERSION/ -Duseshrplib -j "$(nproc)" && \
    curl -sfLO http://apache.org/dist/perl/mod_perl-$MOD_PERL_VERSION.tar.gz && \
    echo "ca2a9e18cdf90f9c6023e786369d5ba75e8dac292ebfea9900c29bf42dc16f74 *mod_perl-$MOD_PERL_VERSION.tar.gz" | sha256sum -c && \
    tar xzf mod_perl-$MOD_PERL_VERSION.tar.gz && \
    cd mod_perl-$MOD_PERL_VERSION && \
    /opt/perl-$PERL_VERSION/bin/perl Makefile.PL MP_NO_THREADS=1 && \
    make -j "$(nproc)" && \
    make install && \
    cd .. && \
    rm -rf mod_perl-$MOD_PERL_VERSION mod_perl-$MOD_PERL_VERSION.tar.gz && \
    apt-get autoremove -yq && \
    rm -rf /var/lib/apt/lists/*

RUN cpan App::cpanminus && \
    cpanm Apache2::AuthCookie && \
    cpanm Apache2::AuthCookieDBI && \
    rm -rf /tmp/* /root/.cpan /root/.cpanm && \
    mkdir /var/tmp/Apache-AuthCookieDBI

COPY Apache-AuthCookieDBI/ /var/tmp/Apache-AuthCookieDBI

RUN cd /var/tmp/Apache-AuthCookieDBI/ && \
    perl Build.PL && \
    ./Build test

WORKDIR /var/tmp/Apache-AuthCookieDBI
```

matisse commented:
That's interesting about the constants. I'm going to play with the Dockerfile (thank you for that - maybe it should be added to this repo?) and see if I can understand what's happening.

matisse commented:
Hmmm. I just tried the Dockerfile and did not get the error messages:

````Step 8/9 : RUN cd /var/tmp/Apache-AuthCookieDBI/ &&     perl Build.PL &&     ./Build test
 ---> Running in 3d7b416c7eae
Created MYMETA.yml and MYMETA.json
Creating new 'Build' script for 'Apache2-AuthCookieDBI' version '2.19'
t/basic.t .. ok
t/utils.t .. ok
All tests successful.
Files=2, Tests=72,  0 wallclock secs ( 0.01 usr  0.01 sys +  0.11 cusr  0.01 csys =  0.14 CPU)
Result: PASS````

esabol commented:
> I'm going to play with the Dockerfile (thank you for that - maybe it should be added to this repo?)

Well, if you want me to add it, I can...

> Hmmm. I just tried the Dockerfile and did not get the error messages:

Did you add a test for `Apache2_4::AuthCookieDBI`? I copied `basic.t` to `apache24.t` and then changed `Apache2::AuthCookieDBI` to `Apache2_4::AuthCookieDBI` in the `use_ok()` test. Or you an just add it to `basic.t`.

By the way, https://metacpan.org/pod/Test::More#use_ok recommends that `use_ok()` tests should be inside a `BEGIN` block, like this:
```Perl
BEGIN { use_ok('Apache2::AuthCookieDBI'); }
```
It probably doesn't make much difference, but let me know if you want me make that change to `basic.t` in this PR.

matisse commented:
Ahh, I see I had not added the Apache2_4 version to the test. I can reproduce the issue.

Interestingly I found his: <https://perl.apache.org/docs/2.0/user/porting/compat.html#mod_perl_1_0_and_2_0_Constants_Coexistence> which describes this problem but in the context of mod_perl 1 and 2 constants.

So I started playing around inside the docker container. I made a copy of `basic.t` name `t/Apache2_4-AuthCookieDBI.t` that tested only `Apache2_4:: AuthCookieDBI` and ran the failing test in the debugger, and saw the the `Apache2::Const` that was loaded was:

   'Apache2/Const.pm' => '/var/tmp/Apache-AuthCookieDBI/t/mock_libs/Apache2/Const.pm'`

Ah ha! So I removed the `use lib "$Bin/mock_libs";` from `t/Apache2_4-AuthCookieDBI.t` and the test pass.

I think I added the mock_libs back when i was trying to create a set of unit tests that were "true" unit tests - did not rely on a server, nor even the local installation of the mod_perl modules.

Anyway, get rid of or comment out `use lib "$Bin/mock_libs";` and we are good.

esabol commented:
> Anyway, get rid of or comment out use lib "$Bin/mock_libs"; and we are good.

I don't think it's that simple? I mean it will pass for me and it'll pass in the above Docker container, but I don't that will work for people who have Apache 2.2.x installed, and I'm guessing it won't work for cpantesters....

esabol commented:
What do you think about adding
```Perl
sub AUTHZ_GRANTED        {2401}
sub AUTHZ_DENIED         {2402}
sub AUTHZ_DENIED_NO_USER {2403}
sub AUTHZ_GENERAL_ERROR  {2404}
```
to `t/mock_libs/Apache2/Const.pm`? Those values are bogus, but I think all that matters is that they are distinct from other values. The simple `use_ok('Apache2_4::AuthCookieDBI');` test passes then. It might even pass on systems that have Apache 2.2.x installed (instead of 2.4.x)? That's probably not a bad thing.... Yes, I just compiled a Docker image with Apache 2.2.34 and mod_perl 2.0.11 and confirmed that.

I've now copied `utils.t` to `apache24.t` and modified it to test `Apache2_4::AuthCookieDBI`. I'll commit these changes soon in the next day or two.

esabol committed:
Add tests for Apache2_4::AuthCookieDBI

esabol commented:
Committed tests. How does it look?

Do you want me to commit the Dockerfiles?

Does the "provides" section of `META.yml` file need to be updated?

esabol committed:
Removed unused Apache2::Consts from t/util24.t

matisse commented:
I think adding the Dockerfile somewhere (not necessarily at the root of the distro) would be helpful, if you don't mind go ahead and do it but I can do it just before I submit to CPAN.

As for `META.yml`  that can be updated easily with `perl Build.PL && ./Build distmeta` (likewise the MANIFEST gets updated with `./Build manifest`) I can do this also so don't worry about it.

Let me know if you want to make other changes, otherwise I'll merge and finish up.

esabol commented:
I’ve made some improvements to the Dockerfile since posting the above version, so I’ll commit it... I was going to create a `docker` subdirectory in the root of the distro and put the files in there, like this:
```
docker/httpd-2.2/Dockerfile
docker/httpd-2.2/Makefile
docker/httpd-2.4/Dockerfile
docker/httpd-2.4/Makefile
```
Or I can put them under `t/` since one could say they are related to testing?
```
t/docker/httpd-2.2/Dockerfile
t/docker/httpd-2.2/Makefile
t/docker/httpd-2.4/Dockerfile
t/docker/httpd-2.4/Makefile
```
Or do you prefer somewhere else?

In case you were wondering, the Makefiles in these directories are there so that one can just “make image” and “make test” and they handle all the Docker syntax for building the image and running the container.

matisse commented:
I think the top level `docker` directory is better - I'm thinking it'll be more obvious (I may add a README inside it) and easier for a developer to use.

esabol committed:
Added Docker files used for testing

esabol commented:
Committed Dockerfiles. I think we are done unless you can think of any other changes you want. Feel free to merge. Thanks for your support!

matisse reviewed (approved):
Thank you Ed I'm going to merge and submit to CPAN probably this weekend.
